### PR TITLE
Updates to GreedySearch/BeamSearch

### PR DESCRIPTION
**Pull request by hariharans29**
### Description
Kernel changes to be able to run models generated using new option introduced in https://github.com/microsoft/onnxruntime/pull/13914

### Motivation and Context
Alternate design for optimization in https://github.com/microsoft/onnxruntime/pull/13809

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc [hunk 1]
```diff
@@ -87,7 +96,17 @@ Status BeamSearch::SetupSubgraphExecutionInfo(const SessionState& session_state,
...
                                         gpt_subgraph_->num_layers);
+    } else if (attribute_name == "init_decoder") {
+      ORT_ENFORCE(init_run_gpt_subgraph_ == nullptr, "SetupSubgraphExecutionInfo should only be called once for each subgraph.");
+      init_run_gpt_subgraph_ = std::make_unique<GptSubgraph>(node, attribute_name, subgraph_session_state.GetGraphViewer());
+      ORT_RETURN_IF_ERROR(init_run_gpt_subgraph_->Setup(session_state, subgraph_session_state));
+      init_run_decoder_feeds_fetches_manager_ = init_run_gpt_subgraph_->GetFeedsFetchesManager();
+      parameters_.SetSubgraphParameters(init_run_gpt_subgraph_->vocab_size,
+                                        init_run_gpt_subgraph_->num_heads,
+                                        init_run_gpt_subgraph_->head_size,
+                                        init_run_gpt_subgraph_->num_layers);
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:107]: nit: extract this as a function

Comment (yufenglee) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:107]: And all subgraph setting up below and above can share the code

**Review by yufenglee — DISMISSED** [review]
:shipit:

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 2]
```diff
@@ -1044,6 +1044,10 @@ ONNX_MS_OPERATOR_SET_SCHEMA(BeamSearch, 1,
                                 .Attr("early_stopping", "early stop or not", AttributeProto::INT, static_cast<int64_t>(0))
                                 .Attr("model_type", "model type: 0 for GPT-2; 1 for encoder decoder like T5", AttributeProto::INT, static_cast<int64_t>(0))
                                 .Attr("encoder", "The subgraph for initialization of encoder and decoder. It will be called once before decoder subgraph.", AttributeProto::GRAPH, OPTIONAL_VALUE)
+                                .Attr("init_decoder",
```
Comment (wangyems) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1047]: maybe we can leverage the encoder_decoderinit so there's no need to add another attribute?

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 3]
```diff
@@ -1085,7 +1089,11 @@ ONNX_MS_OPERATOR_SET_SCHEMA(GreedySearch, 1,
                                 .Attr("decoder_start_token_id", "The id of the token that indicates decoding starts.", AttributeProto::INT, static_cast<int64_t>(-1))
                                 .Attr("no_repeat_ngram_size", "no repeat ngrams size", AttributeProto::INT, static_cast<int64_t>(0))
                                 .Attr("model_type", "model type: 0 for decoder only like GPT-2; 1 for encoder decoder like Bart", AttributeProto::INT, static_cast<int64_t>(0))
-                                .Attr("encoder", "The subgraph for initialization of encoder and decoder. It will be called once before decoder subgraph.", AttributeProto::GRAPH, OPTIONAL_VALUE)
+                                .Attr("encoder", "The subgraph for initialization of encoder and decoder. It will be called once before `decoder` subgraph.", AttributeProto::GRAPH, OPTIONAL_VALUE)
```
Comment (wangyems) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1092]: doc update

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 2 above)
Comment (hariharans29) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1047]: You mean in the conversion tool or here in the op schema where-in we can just re-use the existing attribute or both ? 
Open to any such change. Not sure if using "encoder" in GPT2 was semantically right. Please advise.

**Review by yufenglee — APPROVED** [review]
:shipit:

**Review by wangyems — APPROVED** [review]

**Pull request merged**